packages/circuit-ui/hooks/useCollapsible/useCollapsible.ts

```diff
@@ -0,0 +1,128 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useRef, useCallback, RefObject } from 'react';
+
+import { ClickEvent } from '../../types/events';
+import { uniqueId } from '../../util/id';
+import { useAnimation } from '../useAnimation';
+
+const DEFAULT_HEIGHT = 'auto';
+
+export type CollapsibleOptions = {
+  initialOpen?: boolean;
+  duration?: number;
+  id?: string;
+};
+
+type ButtonProps = {
+  'onClick': (event: ClickEvent) => void;
+  'tabIndex': number;
+  'aria-controls': string;
+  'aria-expanded': 'true' | 'false';
+};
+
+type ContentProps<T> = {
+  'ref': RefObject<T>;
+  'id': string;
+  'style': {
+    overflow: 'hidden';
+    willChange: 'height';
+    opacity: 1 | 0;
+    height: string | 0;
+  };
+  'aria-hidden': null | 'true';
+};
+
+type Collapsible<T> = {
+  isOpen: boolean;
+  toggleOpen: () => void;
+  getButtonProps: (props?: {
+    onClick?: (event: ClickEvent) => void;
+  }) => ButtonProps;
+  getContentProps: (props?: {
+    style?: Record<string, string>;
+  }) => ContentProps<T>;
+};
+
+/**
+ * A hook to build accessible and smoothly animated collapsible sections.
+ * Based on https://inclusive-components.design/collapsible-sections/
+ */
+export function useCollapsible<T extends HTMLElement = HTMLElement>({
+  initialOpen = false,
+  duration = 300,
+  id,
+}: CollapsibleOptions = {}): Collapsible<T> {
+  const contentId = id || uniqueId('collapsible_');
+  const contentElement = useRef<T>(null);
+  const [isOpen, setOpen] = useState(initialOpen);
+  const [height, setHeight] = useState(getHeight(contentElement));
+  const [, setAnimating] = useAnimation();
+  const toggleOpen = useCallback(() => {
+    setAnimating({
+      duration,
+      onStart: () => {
+        setHeight(getHeight(contentElement));
+        if (!isOpen) {
+          setOpen(true);
+        }
+      },
+      onEnd: () => {
+        if (isOpen) {
+          setOpen(false);
+        }
+        setHeight(DEFAULT_HEIGHT);
+      },
+    });
+  }, [isOpen, setAnimating, duration]);
+
+  return {
+    isOpen,
+    toggleOpen,
+    getButtonProps: (props = {}) => ({
+      'onClick': (event: ClickEvent) => {
+        if (props.onClick) {
+          props.onClick(event);
+        }
+        toggleOpen();
+      },
+      'tabIndex': 0,
+      'aria-controls': contentId,
+      'aria-expanded': isOpen ? 'true' : 'false',
+    }),
+    getContentProps: (props = {}) => ({
+      'ref': contentElement,
+      'id': contentId,
+      'style': {
+        overflow: 'hidden',
+        willChange: 'height',
+        opacity: isOpen ? 1 : 0,
+        height: isOpen ? height : 0,
```

Comment:
Animating the `height` property of the element will trigger layout calculation, worth keeping in mind to refactor if we see any performance issues!
https://www.html5rocks.com/en/tutorials/speed/high-performance-animations/

Reply:
👀 what do you think would be a better alternative?

Reply:
I have the same question as Robin: what alternative do you have in mind?

The way I see it, the layout calculations are impossible to avoid since changing the dimensions of the collapsible section inevitably affects the position of the elements below it.

Reply:
Animating with the `transform` property is the most performant way to achieve this, so in this case we would use `scaleY`.
https://developer.mozilla.org/en-US/docs/Web/Performance/Animation_performance_and_frame_rate#css_property_cost
We could use `willChange` css property to promote this component into it's own layer so the animation is optimized, we let the browser know this element will be changing dimensions. Although it should be used scarcely in this case it would help.
 @connor-baer That's true, it changes the positions of the elements below (or on the side), but this way not all of the calculations will need to be done as the browser will have more information on what kind of changes are going to happen! :)
https://developer.mozilla.org/en-US/docs/Web/CSS/will-change

Reply:
`scaleY` doesn't change the space that an element uses, i.e. it doesn't affect the layout:

https://user-images.githubusercontent.com/11017722/128519956-77c61806-7876-41fc-9576-5cae6dc7028e.mp4

That's exactly my point: we _want_ to affect the layout, so `transform` doesn't work in this case.

`will-change` is already used: https://github.com/sumup-oss/circuit-ui/pull/1089/files#diff-952fc020b0d87ecb69db3a5a62e3334079d9ab8c001d54f4bb03d50b7ee58272R111

Reply:
That's very true, we do want to make space to accommodate the collapsed content!
Agreed, it seems best to keep the solution with dynamic height in this case!
Thanks for looking into it!